File: Ejercicios/Ejercicio6/5.cpp
#include <iostream>
using namespace std;

// Interfaz para el envio de mensajes
class Mensaje 
{
public:
    virtual void enviar() = 0;
};

// Clase concreta para enviar mensajes por correo electronico
class CorreoElectronico : public Mensaje 
{
public:
    void enviar() override 
    {
        cout << "Mensaje enviado por correo electronico" << endl;
    }
};

// Clase concreta para enviar mensajes por SMS
class SMS : public Mensaje 
{
public:
    void enviar() override 
    {
        cout << "Mensaje enviado por SMS" << endl;
    }
};

// Clase de alto nivel que depende de la interfaz Mensaje
class Notificador 
{
private:
    Mensaje* mensaje;

public:
    Notificador(Mensaje* mensaje) : mensaje(mensaje) {}

    void enviarMensaje() 
    {
        mensaje->enviar();
    }
};

int main() 
{
    Mensaje* correo = new CorreoElectronico();
    Notificador notificadorCorreo(correo);
    notificadorCorreo.enviarMensaje();

    Mensaje* sms = new SMS();
    Notificador notificadorSMS(sms);
    notificadorSMS.enviarMensaje();

    delete correo;
    delete sms;

    return 0;
}

File: Ejercicios/Ejercicio6/5.py
# Interfaz para el envio de mensajes
class Mensaje:
    def enviar(self):
        pass

# Clase concreta para enviar mensajes por correo electronico
class CorreoElectronico(Mensaje):
    def enviar(self):
        print("Mensaje enviado por correo electronico")

# Clase concreta para enviar mensajes por SMS
class SMS(Mensaje):
    def enviar(self):
        print("Mensaje enviado por SMS")

# Clase de alto nivel que depende de la interfaz Mensaje
class Notificador:
    def __init__(self, mensaje):
        self.mensaje = mensaje

    def enviarmensaje(self):
        self.mensaje.enviar()

correo = CorreoElectronico()
notificadorcorreo = Notificador(correo)
notificadorcorreo.enviamensaje()

sms = SMS()
notificadorsms = Notificador(sms)
notificadorsms.enviarmensaje()
